### ISSUE TITLE
Expose Grid Api and columnApi to JS

### ISSUE THREAD

**Issue by BSd3v** [plotly/dash-ag-grid#194]
Good day, 

I'd like to request a feature to allow for JS to be used seamlessly with the grid's `api` and `columnApi`, this would allow for clientside callback's to work arbitrarily and alter things about the grid's state without need to create specific endpoints in the grid's componet.

The reasoning:

There are tons of features built in to the [api](https://www.ag-grid.com/react-data-grid/grid-api/) and the [columnApi](https://www.ag-grid.com/react-data-grid/column-api/) that to include them all would bloat the component. Exposing these would allow for developers to customize their workflow in an indefinite number of ways.

Some current examples of things that would be available:
[scroll to row/column](https://community.plotly.com/t/ag-grid-vertically-scrolls-the-grid/75457/2)
[drag between grids](https://community.plotly.com/t/dash-ag-grid-how-to-drag-drop-between-grids/75214/8) - would more align with the AG Grid [example](https://www.ag-grid.com/react-data-grid/drag-and-drop/#dragging-between-grids)
[adjusting row height](https://www.ag-grid.com/react-data-grid/row-height/#apiresetrowheights)
More complex workflows for components and editors to interact with other things about the grid.

Complications:

With great power comes great responsibility. There is the chance that updates to the grid may not make their way into the `columnState` in Dash, `rowData` or other changes may also get out of sync.

Possible JS workflow:

```
app.clientside_callback(
""" function (i) {
var gridState = document.getElementById("#grid").gridState()
return window.dash_clientside.no_update
}""",
Output('grid','id'), Input('grid','id'))

app.clientside_callback("""
function resetHeight(n) {
    if (n) {
        gridState.api.resetRowHeights()
    }
    return window.dash_clientside.no_update
}""",
Output('resetHeight','n_clicks'), Input('resetHeight','n_clicks')
```

**Comment by AnnMarieW** [comment]
I agree - being able to access the grid's API from a callback would be very helpful.  Here is a partial list of things that rely on the API but are not yet part of Dash



[Events](https://www.ag-grid.com/react-data-grid/grid-events/)
We currently have `cellClicked` and now `cellDoublclicked` but there are dozens of others.  Some of the items included on the Events page in the AG Grid docs are available via other dash props, but it's not complete.


[Excel Export](https://www.ag-grid.com/react-data-grid/excel-export-api/)
It's possible to use pandas and `dcc.Download` to export data to Excel, however the Excel Export API has an extensive API that is customized to work with the grid features.  Currently, you can download to Excel  from the Enterprise context menu, but it would add flexibility to be able access the API in a callback. There are also [other API available](https://www.ag-grid.com/react-data-grid/excel-export-api/#api-methods) not accessible from the context menu.


[Range Selection](https://www.ag-grid.com/react-data-grid/range-selection/#range-selection-api)
This is  available in DataTable and is an AG Grid Enterprise feature.  It's possible to enable Range Selection, but you cannot yet interact with the selected data in a Dash callback.


This [Advanced Range Selection Example](https://www.ag-grid.com/react-data-grid/range-selection/#example-advanced-range-selection) requires `rangeSelectionChanged` event and `processCellForClipboard` and `processCellFromClipboard` to accept functions.


[Context menu](https://www.ag-grid.com/react-data-grid/context-menu/)
It's possible to enable the context menu in Dash, but it's not possible to customize the  menu.  See this [Context Menu example](https://www.ag-grid.com/react-data-grid/context-menu/#context-menu-example).  This was requested recently on the [forum](https://community.plotly.com/t/context-menu-in-dash-ag-grid/75674)

I'll continue to add to this post as I work on adding more Enterprise examples to the docs.

**Comment by sdidier-dev** [comment]
As I worked on the Column State docs, one more function could be useful: `applyColumnState` to make a [partial state update](https://www.ag-grid.com/react-data-grid/column-state/#partial-state).
For now, setting `columnState` :
- we have to provide the full column set, even if there is no state changes
- the column order of the new state is also applied, even if we don't want to.

Using `applyColumnState` we could easily:
- provide only the columns to update, not the full set
- eventually provide default values
- choose to keep the current column order or apply the order of the new state

**Comment by BSd3v** [comment]
> As I worked on the Column State docs, one more function could be useful: `applyColumnState` to make a [partial state update](https://www.ag-grid.com/react-data-grid/column-state/#partial-state). 

For clarification of this topic, this function would be reachable under the `columnApi`.

**Comment by BSd3v** [comment]
When using a changelog, similar to what I have posted https://github.com/plotly/dash-ag-grid/issues/190, reloading the info is easy if you only able to make edits to 1 column, replaying the data back is as simple as passing `reloadData` inside the update option of `rowTransaction`.

The issue arises when you allow for multiple columns to be edited at the same time. This means that in order to efficiently update the correct info, we need to match the `rowIndex` and pull in the current `rowData` from that index. We can iterate through to find the details, which can be troublesome because we need to parse the data and make sure it lines up with the `rowIndex`.

Allowing for the api means that this task becomes a call to pull the node based upon the `rowIndex`, an exact match that we can then update the desired key and send it back to the grid. Much easier to solve and no looping or parsing of data.

**Comment by alexcjohnson** [comment]
Alright, I'm sold. We should make clear though that this API is maintained by AG Grid and as such is subject to potentially breaking changes any time we upgrade AG Grid, even in a dash-ag-grid minor update.

An idea how to access this: `dash_ag_grid.getApi(id)` and `dash_ag_grid.getColumnApi(id)` ie attach these functions to the global exported namespace so users need not interact directly with the DOM - particularly useful if `id` is a PMC object, we can reuse [`stringifyId`](https://github.com/plotly/dash-ag-grid/blob/89af6a8aa904324a74f1a0ecbf5d1f6814168ab4/src/lib/fragments/AgGrid.react.js#L97) that we added for aligned grids.

**Comment by BSd3v** [comment]
Hello @plbremer,

There will not be a way directly to update props, but will be able to pull the info upon some interaction via a clientside_callback.
